# Esercizio fine modulo M4

In [1]:
import pandas as pd
import numpy as np
dati = pd.read_csv("owid-covid-data.csv", delimiter=",", low_memory=False)

In [2]:
dati.head()

iso_code continent     location        date  total_cases  new_cases  \
0      AFG      Asia  Afghanistan  2020-01-03          NaN        0.0   
1      AFG      Asia  Afghanistan  2020-01-04          NaN        0.0   
2      AFG      Asia  Afghanistan  2020-01-05          NaN        0.0   
3      AFG      Asia  Afghanistan  2020-01-06          NaN        0.0   
4      AFG      Asia  Afghanistan  2020-01-07          NaN        0.0   

   new_cases_smoothed  total_deaths  new_deaths  new_deaths_smoothed  ...  \
0                 NaN           NaN         0.0                  NaN  ...   
1                 NaN           NaN         0.0                  NaN  ...   
2                 NaN           NaN         0.0                  NaN  ...   
3                 NaN           NaN         0.0                  NaN  ...   
4                 NaN           NaN         0.0                  NaN  ...   

   male_smokers  handwashing_facilities  hospital_beds_per_thousand  \
0           NaN                  37.746                         0.5   
1           NaN                  37.746                         0.5   
2           NaN                  37.746                         0.5   
3           NaN                  37.746                         0.5   
4           NaN                  37.746                         0.5   

   life_expectancy  human_development_index  population  \
0            64.83                    0.511  41128772.0   
1            64.83                    0.511  41128772.0   
2            64.83                    0.511  41128772.0   
3            64.83                    0.511  41128772.0   
4            64.83                    0.511  41128772.0   

   excess_mortality_cumulative_absolute  excess_mortality_cumulative  \
0                                   NaN                          NaN   
1                                   NaN                          NaN   
2                                   NaN                          NaN   
3                                   NaN                          NaN   
4                                   NaN                          NaN   

   excess_mortality  excess_mortality_cumulative_per_million  
0               NaN                                      NaN  
1               NaN                                      NaN  
2               NaN                                      NaN  
3               NaN                                      NaN  
4               NaN                                      NaN  

[5 rows x 67 columns]

## 1) Si richiede di verificare le dimensioni del dataset e le diciture presenti nell'intestazione:

In [23]:
dati.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 311365 entries, 0 to 311364
Data columns (total 67 columns):
 #   Column                                      Non-Null Count   Dtype  
---  ------                                      --------------   -----  
 0   iso_code                                    311365 non-null  object 
 1   continent                                   296585 non-null  object 
 2   location                                    311365 non-null  object 
 3   date                                        311365 non-null  object 
 4   total_cases                                 275335 non-null  float64
 5   new_cases                                   302655 non-null  float64
 6   new_cases_smoothed                          301391 non-null  float64
 7   total_deaths                                254635 non-null  float64
 8   new_deaths                                  302735 non-null  float64
 9   new_deaths_smoothed                         301505 non-null  float64
 

## 2) Si chiede poi per ogni continente di trovare il numero di casi totali avvenuti in quello stesso continente; si chiedi di non considerare eventuali locazioni che nel dataset non appartengono ad alcun continente:

In [50]:
dati.groupby(by = "continent", as_index = False).agg(casi_totali = ("new_cases", np.sum))

continent  casi_totali
0         Africa   13088991.0
1           Asia  297234406.0
2         Europe  249319188.0
3  North America  124175560.0
4        Oceania   14073975.0
5  South America   68568119.0

## 3) Sempre riguardo i casi di COVID totali, si chiede di sviluppare una funzione che prenda in input il dataset e due nomi di continenti, e che ne confronti i seguenti relativi descrittori statistici: valori minimo e massimo, media, e percentuale rispetto al numero di casi totali nel mondo (in questo caso calcolati anche sulle locazioni senza indicazione di continente):

In [22]:
def funzione1(continent1, continent2):

    dati_continent = dati.loc[:, "continent":"new_cases":4]

    dati_continent1_filtro = dati_continent[dati_continent["continent"]==continent1]
    dati_continent1 = dati_continent1_filtro[dati_continent1_filtro["new_cases"]!=0]
    dati_continent1_cases = dati_continent1.loc[:, "new_cases"]
    dati_continent1_min_cases = dati_continent1_cases.min()
    dati_continent1_max_cases = dati_continent1_cases.max()
    dati_continent1_avg = dati_continent1_cases.mean()

    dati_continent2_filtro = dati_continent[dati_continent["continent"]==continent2]
    dati_continent2 = dati_continent2_filtro[dati_continent2_filtro["new_cases"]!=0]
    dati_continent2_cases = dati_continent2.loc[:, "new_cases"]
    dati_continent2_min_cases = dati_continent2_cases.min()
    dati_continent2_max_cases = dati_continent2_cases.max()
    dati_continent2_avg = dati_continent2_cases.mean()
    
    confronto_continenti = dati.groupby(by = "continent", as_index = False).agg(casi_totali = ("new_cases", np.sum))

    cf_continenti = dati.groupby(by = "continent", as_index = True).agg(casi_totali = ("new_cases", np.sum))

    numeri_confronto = confronto_continenti.loc[:, "casi_totali"].astype(int)
    totale_continent1 = cf_continenti.loc[continent1, :]
    totale_continent2 = cf_continenti.loc[continent2, :]
    totale_casi = numeri_confronto.sum()
    perc_continent1 = (totale_continent1 * 100)/totale_casi
    perc_continent2 = (totale_continent2 * 100)/totale_casi

    differenza_min = dati_continent1_min_cases - dati_continent2_min_cases
    differenza_max = dati_continent1_max_cases - dati_continent2_max_cases
    differenza_avg = dati_continent1_avg - dati_continent2_avg
    differenza_perc = perc_continent1 - perc_continent2

    confronto_dati = {"Min": [dati_continent1_min_cases, dati_continent2_min_cases, differenza_min], "Max": [dati_continent1_max_cases, dati_continent2_max_cases, differenza_max], "Media": [dati_continent1_avg, dati_continent2_avg, differenza_avg], "% su Tot": [perc_continent1, perc_continent2, differenza_perc]}
    confronto = pd.DataFrame(data=confronto_dati, index = [continent1, continent2, "Confronto"])
    print(confronto.round(0).astype(int))

## 4) Si chiede poi di effettuare lo stesso tipo di analisi - anche in questo caso sviluppando una funzione ad hoc - per il numero di vaccinazioni totali per ogni continente:

In [57]:
dati.groupby(by = "continent", as_index = False).agg(vaccinazioni_totali = ("new_vaccinations", np.sum))

continent  vaccinazioni_totali
0         Africa         1.173671e+08
1           Asia         7.547783e+09
2         Europe         1.264676e+09
3  North America         9.974152e+08
4        Oceania         7.454216e+07
5  South America         8.430426e+08

In [3]:
def funzione2(continent1, continent2):

    dati_continent = dati.loc[:, "continent":"new_vaccinations":37]

    dati_continent1_filtro = dati_continent[dati_continent["continent"]==continent1]
    dati_continent1 = dati_continent1_filtro[dati_continent1_filtro["new_vaccinations"]!=0]
    dati_continent1_vaccinations = dati_continent1.loc[:, "new_vaccinations"]
    dati_continent1_min_vaccinations = dati_continent1_vaccinations.min()
    dati_continent1_max_vaccinations = dati_continent1_vaccinations.max()
    dati_continent1_avg = dati_continent1_vaccinations.mean()

    dati_continent2_filtro = dati_continent[dati_continent["continent"]==continent2]
    dati_continent2 = dati_continent2_filtro[dati_continent2_filtro["new_vaccinations"]!=0]
    dati_continent2_vaccinations = dati_continent2.loc[:, "new_vaccinations"]
    dati_continent2_min_vaccinations = dati_continent2_vaccinations.min()
    dati_continent2_max_vaccinations = dati_continent2_vaccinations.max()
    dati_continent2_avg = dati_continent2_vaccinations.mean()

    confronto_continente = dati.groupby(by = "continent", as_index = False).agg(vaccinations_totali = ("new_vaccinations", np.sum))

    cf_continenti = dati.groupby(by = "continent", as_index = True).agg(vaccinations_totali = ("new_vaccinations", np.sum))

    numeri_confronto = confronto_continente.loc[:, "vaccinations_totali"]
    totale_continent1 = cf_continenti.loc[continent1, "vaccinations_totali"]
    totale_continent2 = cf_continenti.loc[continent2, "vaccinations_totali"]
    totale_vaccinations = numeri_confronto.sum()
    perc_continent1 = (totale_continent1 * 100)/totale_vaccinations
    perc_continent2 = (totale_continent2 * 100)/totale_vaccinations

    differenza_min = dati_continent1_min_vaccinations - dati_continent2_min_vaccinations
    differenza_max = dati_continent1_max_vaccinations - dati_continent2_max_vaccinations
    differenza_avg = dati_continent1_avg - dati_continent2_avg
    differenza_perc = perc_continent1 - perc_continent2

    confronto_dati = {"Min": [dati_continent1_min_vaccinations, dati_continent2_min_vaccinations, differenza_min], "Max": [dati_continent1_max_vaccinations, dati_continent2_max_vaccinations, differenza_max], "Media": [dati_continent1_avg, dati_continent2_avg, differenza_avg], "% su Tot": [perc_continent1, perc_continent2, differenza_perc]}
    confronto = pd.DataFrame(data=confronto_dati, index = [continent1, continent2, "Confronto"])
    print(confronto.round(0).astype(int))

## 5) Alla fine, basandosi sui calcoli fatti, il committente chiede di stilare un breve (tre o quattro righe) paragrafo testuale riassuntivo sulle statistiche di casi e vaccinazioni, che si concentri solo sulle differenze esistenti tra Europa, Sud America e Oceania:

In [20]:
def confronto_casi_covid(continent1, continent2, continent3):
    
    dati_continent = dati.loc[:, "continent":"new_cases":4]

    dati_continent1_filtro = dati_continent[dati_continent["continent"]==continent1]
    dati_continent1 = dati_continent1_filtro[dati_continent1_filtro["new_cases"]!=0]
    dati_continent1_cases = dati_continent1.loc[:, "new_cases"]
    dati_continent1_min_cases = dati_continent1_cases.min()
    dati_continent1_max_cases = dati_continent1_cases.max()
    dati_continent1_avg = dati_continent1_cases.mean()

    dati_continent2_filtro = dati_continent[dati_continent["continent"]==continent2]
    dati_continent2 = dati_continent2_filtro[dati_continent2_filtro["new_cases"]!=0]
    dati_continent2_cases = dati_continent2.loc[:, "new_cases"]
    dati_continent2_min_cases = dati_continent2_cases.min()
    dati_continent2_max_cases = dati_continent2_cases.max()
    dati_continent2_avg = dati_continent2_cases.mean()
    
    dati_continent3_filtro = dati_continent[dati_continent["continent"]==continent3]
    dati_continent3 = dati_continent3_filtro[dati_continent3_filtro["new_cases"]!=0]
    dati_continent3_cases = dati_continent3.loc[:, "new_cases"]
    dati_continent3_min_cases = dati_continent3_cases.min()
    dati_continent3_max_cases = dati_continent3_cases.max()
    dati_continent3_avg = dati_continent3_cases.mean()
    
    confronto_continenti = dati.groupby(by = "continent", as_index = False).agg(casi_totali = ("new_cases", np.sum))

    cf_continenti = dati.groupby(by = "continent", as_index = True).agg(casi_totali = ("new_cases", np.sum))

    numeri_confronto = confronto_continenti.loc[:, "casi_totali"].astype(int)
    totale_continent1 = cf_continenti.loc[continent1, :]
    totale_continent2 = cf_continenti.loc[continent2, :]
    totale_continent3 = cf_continenti.loc[continent3, :]
    totale_casi = numeri_confronto.sum()
    perc_continent1 = (totale_continent1 * 100)/totale_casi
    perc_continent2 = (totale_continent2 * 100)/totale_casi
    perc_continent3 = (totale_continent3 * 100)/totale_casi    

    confronto_dati = {"Min": [dati_continent1_min_cases, dati_continent2_min_cases, dati_continent3_min_cases], "Max": [dati_continent1_max_cases, dati_continent2_max_cases, dati_continent3_max_cases], "Media": [dati_continent1_avg, dati_continent2_avg, dati_continent3_avg], "% su Tot": [perc_continent1, perc_continent2, perc_continent3]}
    confronto = pd.DataFrame(data=confronto_dati, index = [continent1, continent2, continent3])
    print(confronto.round(0).astype(int))

In [21]:
def confronto_vaccinazioni(continent1, continent2, continent3):

    dati_continent = dati.loc[:, "continent":"new_vaccinations":37]

    dati_continent1_filtro = dati_continent[dati_continent["continent"]==continent1]
    dati_continent1 = dati_continent1_filtro[dati_continent1_filtro["new_vaccinations"]!=0]
    dati_continent1_vaccinations = dati_continent1.loc[:, "new_vaccinations"]
    dati_continent1_min_vaccinations = dati_continent1_vaccinations.min()
    dati_continent1_max_vaccinations = dati_continent1_vaccinations.max()
    dati_continent1_avg = dati_continent1_vaccinations.mean()

    dati_continent2_filtro = dati_continent[dati_continent["continent"]==continent2]
    dati_continent2 = dati_continent2_filtro[dati_continent2_filtro["new_vaccinations"]!=0]
    dati_continent2_vaccinations = dati_continent2.loc[:, "new_vaccinations"]
    dati_continent2_min_vaccinations = dati_continent2_vaccinations.min()
    dati_continent2_max_vaccinations = dati_continent2_vaccinations.max()
    dati_continent2_avg = dati_continent2_vaccinations.mean()
    
    dati_continent3_filtro = dati_continent[dati_continent["continent"]==continent3]
    dati_continent3 = dati_continent3_filtro[dati_continent3_filtro["new_vaccinations"]!=0]
    dati_continent3_vaccinations = dati_continent3.loc[:, "new_vaccinations"]
    dati_continent3_min_vaccinations = dati_continent3_vaccinations.min()
    dati_continent3_max_vaccinations = dati_continent3_vaccinations.max()
    dati_continent3_avg = dati_continent3_vaccinations.mean()

    confronto_continente = dati.groupby(by = "continent", as_index = False).agg(vaccinations_totali = ("new_vaccinations", np.sum))

    cf_continenti = dati.groupby(by = "continent", as_index = True).agg(vaccinations_totali = ("new_vaccinations", np.sum))

    numeri_confronto = confronto_continente.loc[:, "vaccinations_totali"]
    totale_continent1 = cf_continenti.loc[continent1, "vaccinations_totali"]
    totale_continent2 = cf_continenti.loc[continent2, "vaccinations_totali"]
    totale_continent3 = cf_continenti.loc[continent3, "vaccinations_totali"]
    totale_vaccinations = numeri_confronto.sum()
    perc_continent1 = (totale_continent1 * 100)/totale_vaccinations
    perc_continent2 = (totale_continent2 * 100)/totale_vaccinations
    perc_continent3 = (totale_continent2 * 100)/totale_vaccinations

    confronto_dati = {"Min": [dati_continent1_min_vaccinations, dati_continent2_min_vaccinations, dati_continent3_min_vaccinations], "Max": [dati_continent1_max_vaccinations, dati_continent2_max_vaccinations, dati_continent3_max_vaccinations], "Media": [dati_continent1_avg, dati_continent2_avg, dati_continent3_avg], "% su Tot": [perc_continent1, perc_continent2, perc_continent3]}
    confronto = pd.DataFrame(data=confronto_dati, index = [continent1, continent2, continent3])
    print(confronto.round(0).astype(int))

### Tabella 1 - Confronto casi di covid

In [23]:
confronto_casi_covid("Europe", "South America", "Oceania")

               Min     Max  Media  % su Tot
Europe           1  500563   4978        32
South America    1  298408   5731         8
Oceania          1  588813   2562         1


### Tabella 2 - Confronto vaccinazioni

In [8]:
confronto_vaccinazioni("Europe", "South America", "Oceania")

               Min      Max   Media  % su Tot
Europe           1  2389472   60260        12
South America    1  3976605  150921         8
Oceania          1   351226   56859         8


#### Dalle due tabelle sovrastanti, che racchiudono alcune statistiche riguardanti i casi covid (Tabella 1) e il numero di vaccinazioni (Tabella 2) dei continenti indicati, possiamo notare che:

- L'Oceania è il continente che ha raggiunto il numero più alto di casi giornalieri con 588.813 casi;
- Il numero minimo di casi giornalieri pari a 1 probabilmente è dovuto ai primi giorni dalla scoperta del virus, quando venivano registrati i primi pazienti 0;
- La media di casi giornalieri per il Sud America è la più alta tra quelli presi in esame con 5731 casi giornalieri, mentre l'Oceania la media più bassa, 2562;
- L'Europa detiene invece il numero più alti di casi rispetto al totale mondiale, il 32%;

- Per quanto riguarda le vaccinazioni, l'Europa è il paese con un più alto numero di vaccinati rispetto agli altri due paesi (12%);
- Sempre parlando di vaccinazioni, il Sud America è quello che ha effettuato più vaccinazioni in una giornata (351.226) e ha la media giornaliera più alta (150.951), quasi il triplo rispetto agli altri due continenti presi in esame.